Core/RetroAchievements.cpp

```diff
@@ -122,1839 +62,577 @@ void OSDUpdateBackgroundProgressDialog(const char *str_id, std::string message,
-		{
-			FormattedError("%s failed: empty response and/or status code %d", RAPIStructName<T>(), status_code);
-			LogFailedResponseJSON(data);
-			return;
-		}
-
-		// ensure null termination, rapi needs it
-		data.push_back(0);
-
-		const int error = ParseFunc(this, reinterpret_cast<const char *>(data.data()));
-		initialized = (error == RC_OK);
+	return rc_client_get_hardcore_enabled(g_rcClient);
+}
 
-		const rc_api_response_t &response = static_cast<T *>(this)->response;
-		if (error != RC_OK)
-		{
-			FormattedError("%s failed: parse function returned %d (%s)", RAPIStructName<T>(), error, rc_error_str(error));
-			LogFailedResponseJSON(data);
-		} else if (!response.succeeded)
-		{
-			FormattedError("%s failed: %s", RAPIStructName<T>(),
-				response.error_message ? response.error_message : "<no error>");
-			LogFailedResponseJSON(data);
-		}
+bool WarnUserIfChallengeModeActive(const char *message) {
+	if (!ChallengeModeActive()) {
+		return false;
 	}
 
-	~RAPIResponse()
-	{
-		if (initialized)
-			DestroyFunc(this);
+	const char *showMessage = message;
+	if (!message) {
+		auto ac = GetI18NCategory(I18NCat::ACHIEVEMENTS);
+		showMessage = ac->T("This feature is not available in Challenge Mode");
 	}
 
-	operator bool() const { return initialized && static_cast<const T *>(this)->response.succeeded; }
-};
-
-} // namespace Achievements
-
-void Achievements::FormattedError(const char *format, ...)
-{
-	std::va_list ap;
-	va_start(ap, format);
-	char buffer[1024];
-	vsnprintf(buffer, sizeof(buffer), format, ap);
-	va_end(ap);
-
-	ERROR_LOG(ACHIEVEMENTS, "%s", buffer);
-	// Host::AddOSDMessage(std::move(error), 10.0f);
+	g_OSD.Show(OSDType::MESSAGE_WARNING, showMessage, 3.0f);
+	return true;
 }
 
-void Achievements::LogFailedResponseJSON(const Common::HTTPDownloader::Request::Data &data)
-{
-	const std::string str_data(reinterpret_cast<const char *>(data.data()), data.size());
-	ERROR_LOG(ACHIEVEMENTS, "API call failed. Response JSON was:\n%s", str_data.c_str());
+bool IsBlockingExecution() {
+	return g_isIdentifying;
 }
 
-const Achievements::Achievement *Achievements::GetAchievementByID(u32 id)
-{
-	for (const Achievement &ach : s_achievements)
-	{
-		if (ach.id == id)
-			return &ach;
+static u32 GetGameID() {
+	if (!g_rcClient) {
+		return 0;
 	}
 
-	return nullptr;
-}
-
-Achievements::Achievement *Achievements::GetMutableAchievementByID(u32 id)
-{
-	for (Achievement &ach : s_achievements)
-	{
-		if (ach.id == id)
-			return &ach;
+	const rc_client_game_t *info = rc_client_get_game_info(g_rcClient);
+	if (!info) {
+		return 0;
 	}
+	return info->id;  // 0 if not identified
+}
 
-	return nullptr;
+bool IsActive() {
+	return GetGameID() != 0;
 }
 
-void Achievements::ClearGameInfo(bool clear_achievements, bool clear_leaderboards)
-{
-	const bool had_game = (s_game_id != 0);
+// This is the function the rc_client will use to read memory for the emulator. we don't need it yet,
+// so just provide a dummy function that returns "no memory read".
```

Comment:
Hm, this comment seems to describe an alternate timeline.

-[Unknown]

-[Unknown]

Reply:
Yeah, copied from the integration guide, oops. https://github.com/RetroAchievements/rcheevos/wiki/rc_client-integration